File: sciencefair/matcher.c
/*
 * Regular expression implementation.
 * Supports only ( | ) * + ?.  No escapes.
 * Compiles to NFA and then simulates NFA
 * using Thompson's algorithm.
 *
 * See also http://swtch.com/~rsc/regexp/ and
 * Thompson, Ken.  Regular Expression Search Algorithm,
 * Communications of the ACM 11(6) (June 1968), pp. 419-422.
 *
 * Copyright (c) 2007 Russ Cox.
 * Can be distributed under the MIT license, see bottom of file.
 */

// modified by E nov 2022, all comments starting with // are mine
// original program was to see if a regular expression matches a string,
// modified to search for matching substrings in larger text/file
// original src file: https://swtch.com/~rsc/regexp/nfa.c.txt
// as the code is mostly for educational purposes, russ didnt care about
// memory leaks etc, so neither do i (although there are only few)
//
// compile with optization: cc -O3 matcher.c -o matcher
//
// example usage:
//   ./matcher "data(innbrudd|språk)" no.txt
//   ./matcher " 100* " no.txt
//   echo "hello cat baab dog hello" | ./matcher "hello|ba*b"
// grep equivalents:
//   grep --line-number "data\(innbrudd\|språk\)" no.txt
//   grep --line-number " 100* " no.txt
//   echo "hello cat baab dog hello" | grep --line-number "hello\|ba*b"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*
 * Convert infix regexp re to postfix notation.
 * Insert . as explicit concatenation operator.
 * Cheesy parser, return static buffer.
 */
char*
re2post(char *re)
{
	int nalt, natom;
	static char buf[8000];
	char *dst;
	struct {
		int nalt;
		int natom;
	} paren[100], *p;

	p = paren;
	dst = buf;
	nalt = 0;
	natom = 0;
	if(strlen(re) >= sizeof buf/2)
		return NULL;
	for(; *re; re++){
		switch(*re){
		case '(':
			if(natom > 1){
				--natom;
				*dst++ = '.';
			}
			if(p >= paren+100)
				return NULL;
			p->nalt = nalt;
			p->natom = natom;
			p++;
			nalt = 0;
			natom = 0;
			break;
		case '|':
			if(natom == 0)
				return NULL;
			while(--natom > 0)
				*dst++ = '.';
			nalt++;
			break;
		case ')':
			if(p == paren)
				return NULL;
			if(natom == 0)
				return NULL;
			while(--natom > 0)
				*dst++ = '.';
			for(; nalt > 0; nalt--)
				*dst++ = '|';
			--p;
			nalt = p->nalt;
			natom = p->natom;
			natom++;
			break;
		case '*':
		case '+':
		case '?':
			if(natom == 0)
				return NULL;
			*dst++ = *re;
			break;
		default:
			if(natom > 1){
				--natom;
				*dst++ = '.';
			}
			*dst++ = *re;
			natom++;
			break;
		}
	}
	if(p != paren)
		return NULL;
	while(--natom > 0)
		*dst++ = '.';
	for(; nalt > 0; nalt--)
		*dst++ = '|';
	*dst = 0;
	return buf;
}

/*
 * Represents an NFA state plus zero or one or two arrows exiting.
 * if c == Match, no arrows out; matching state.
 * If c == Split, unlabeled arrows to out and out1 (if != NULL).
 * If c < 256, labeled arrow with character c to out.
 */
enum
{
	Match = 256,
	Split = 257
};
typedef struct State State;
struct State
{
	int c;
	State *out;
	State *out1;
	int lastlist;
	int start;
};
State matchstate = { Match };	/* matching state */
int nstate;

/* Allocate and initialize State */
State*
state(int c, State *out, State *out1)
{
	State *s;

	nstate++;
	s = malloc(sizeof *s);
	s->lastlist = 0;
	s->start = 0;
	s->c = c;
	s->out = out;
	s->out1 = out1;
	return s;
}

/*
 * A partially built NFA without the matching state filled in.
 * Frag.start points at the start state.
 * Frag.out is a list of places that need to be set to the
 * next state for this fragment.
 */
typedef struct Frag Frag;
typedef union Ptrlist Ptrlist;
struct Frag
{
	State *start;
	Ptrlist *out;
};

/* Initialize Frag struct. */
Frag
frag(State *start, Ptrlist *out)
{
	Frag n = { start, out };
	return n;
}

/*
 * Since the out pointers in the list are always
 * uninitialized, we use the pointers themselves
 * as storage for the Ptrlists.
 */
union Ptrlist
{
	Ptrlist *next;
	State *s;
};

/* Create singleton list containing just outp. */
Ptrlist*
list1(State **outp)
{
	Ptrlist *l;

	l = (Ptrlist*)outp;
	l->next = NULL;
	return l;
}

/* Patch the list of states at out to point to start. */
void
patch(Ptrlist *l, State *s)
{
	Ptrlist *next;

	for(; l; l=next){
		next = l->next;
		l->s = s;
	}
}

/* Join the two lists l1 and l2, returning the combination. */
Ptrlist*
append(Ptrlist *l1, Ptrlist *l2)
{
	Ptrlist *oldl1;

	oldl1 = l1;
	while(l1->next)
		l1 = l1->next;
	l1->next = l2;
	return oldl1;
}

/*
 * Convert postfix regular expression to NFA.
 * Return start state.
 */
State*
post2nfa(char *postfix)
{
	char *p;
	Frag stack[1000], *stackp, e1, e2, e;
	State *s;

	// fprintf(stderr, "postfix: %s\n", postfix);

	if(postfix == NULL)
		return NULL;

	#define push(s) *stackp++ = s
	#define pop() *--stackp

	stackp = stack;
	for(p=postfix; *p; p++){
		switch(*p){
		default:
			s = state(*p, NULL, NULL);
			push(frag(s, list1(&s->out)));
			break;
		case '.':	/* catenate */
			e2 = pop();
			e1 = pop();
			patch(e1.out, e2.start);
			push(frag(e1.start, e2.out));
			break;
		case '|':	/* alternate */
			e2 = pop();
			e1 = pop();
			s = state(Split, e1.start, e2.start);
			push(frag(s, append(e1.out, e2.out)));
			break;
		case '?':	/* zero or one */
			e = pop();
			s = state(Split, e.start, NULL);
			push(frag(s, append(e.out, list1(&s->out1))));
			break;
		case '*':	/* zero or more */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(s, list1(&s->out1)));
			break;
		case '+':	/* one or more */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(e.start, list1(&s->out1)));
			break;
		}
	}

	e = pop();
	if(stackp != stack)
		return NULL;

	patch(e.out, &matchstate);
	return e.start;
#undef pop
#undef push
}

typedef struct List List;
struct List
{
	State **s;
	int n;
};
List l1, l2;
static int listid;

void addstate(List*, State*, int);
void step(List*, State*, int, int, List*);

/* Compute initial state list */
List*
startlist(State *start, List *l)
{
	l->n = 0;
	listid++;
	addstate(l, start, 0);
	return l;
}

/* Check whether state list contains a match. */
// int
// ismatch(List *l)
// {
// 	int i;
//
// 	for(i=0; i<l->n; i++)
// 		if(l->s[i] == &matchstate)
// 			return 1;
// 	return 0;
// }
// dont need this: instead check if matchstate.lastlist == listid

/* Add s to l, following unlabeled arrows. */
void
addstate(List *l, State *s, int start)
{
	if(s == NULL || s->lastlist == listid)
		return;
	s->lastlist = listid;
	s->start = start;
	if(s->c == Split){
		/* follow unlabeled arrows */
		addstate(l, s->out, start);
		addstate(l, s->out1, start);
		return;
	}
	l->s[l->n++] = s;
}

/*
 * Step the NFA from the states in clist
 * past the character c,
 * to create next NFA state set nlist.
 */
void
step(List *clist, State *start, int c, int j, List *nlist)
{
	int i;
	State *s;

	addstate(clist, start, j);
	listid++;
	nlist->n = 0;
	for(i=0; i<clist->n; i++){
		s = clist->s[i];
		if(s->c == c)
			addstate(nlist, s->out, s->start);
	}
}

// for storing info on matching substrings
typedef struct Slice Slice;
struct Slice {
	int start;
	int end;
	Slice *next;
};

void
addslice(Slice **slicep, int start, int end)
{
	for(; *slicep; slicep = &(*slicep)->next);
	*slicep = malloc(sizeof **slicep);
	(*slicep)->start = start;
	(*slicep)->end = end;
	(*slicep)->next = NULL;
}

void
match(State *start, char *s, Slice **matchesp)
{
	int i, c;
	List *clist, *nlist, *t;

	clist = startlist(start, &l1);
	nlist = &l2;
	i = 0;
	for(; s[i]; i++){
		// want to include æøå, but possibly introduces bug when
		// it comes to UTF-8 charactars where 256 (our epsilon char) occurs
		// in a byte and its following bit? idk strings are hard
		// but things seem to work as they are now
		// c = s[i] & 0xFF;
		c = s[i];
		step(clist, start, c, i, nlist);
 		t = clist; clist = nlist; nlist = t;	/* swap clist, nlist */
		// if(ismatch(clist))
		if(matchstate.lastlist == listid)
			addslice(matchesp, matchstate.start, i + 1);
	}
}

char *RED, *GREEN, *BLUE, *ENDC;

void
printmatches(Slice *matches, char *line, int linenr)
{
	Slice *slice;
	int i, j, k;

	if(matches == NULL)
		return;

	printf("%s%d%s:%s", GREEN, linenr, BLUE, ENDC);

	k = 0;
	for(slice = matches; slice; slice = slice->next){
		i = slice->start;
		j = slice->end;
		for (; k < i; k++)
			putchar(line[k]);

		fputs(RED, stdout);
		for (; k < j; k++)
			putchar(line[k]);

		fputs(ENDC, stdout);
	}

	for (; line[k]; k++)
		putchar(line[k]);
}

int
main(int argc, char **argv)
{
	char *post;
	State *start;
	int atty;
	FILE *in;
	char *line;
	int linenr;
	size_t len;
	ssize_t read;

 	if(argc < 2){
 		fprintf(stderr, "usage: nfa regexp file\n");
 		fprintf(stderr, "see source file for examples\n");
 		return 1;
 	}

	// this wont pick up all bad regexps,
	// resulting in seg fault sometimes
 	post = re2post(argv[1]);
 	if(post == NULL){
 		fprintf(stderr, "bad regexp %s\n", argv[1]);
 		fprintf(stderr, "see source file for examples\n");
 		return 1;
 	}

 	start = post2nfa(post);
 	if(start == NULL){
 		fprintf(stderr, "error in post2nfa %s\n", post);
 		return 1;
 	}

	atty = isatty(fileno(stdout));
	RED   = atty ? "\033[31m" : "";
	GREEN = atty ? "\033[32m" : "";
	BLUE  = atty ? "\033[34m" : "";
	ENDC  = atty ? "\033[m"   : "";

 	l1.s = malloc(nstate*sizeof l1.s[0]);
 	l2.s = malloc(nstate*sizeof l2.s[0]);
	line = NULL;
	len = 0;
	in = argc > 2 ? fopen(argv[2], "r") : stdin;
	for(linenr = 0; (read = getline(&line, &len, in)) != -1; linenr++){
		Slice *matches = NULL;
		match(start, line, &matches);
		printmatches(matches, line, linenr);
	}

 	return 0;
}

/*
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall
 * be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

File: sciencefair/matcher.py
import fileinput
import sys

# usage: python matcher.py REGEXP FILE
# REGEXP must be quoted since we may use | as an operator
# regular expressions are reverse polish notation with
#   | alternation
#   . concatenation
#   * kleene star
# with no possibility for escaping
#
# example usage:
#   python matcher.py "da.t.a.in.n.b.r.u.d.d.sp.r.å.k.|." no.txt
#   python matcher.py " 1.0.0*. ." no.txt
#   echo "hello cat baab dog hello" | python matcher.py "he.l.l.o.ba*.b.|"
# grep equivalents:
#   grep --line-number "data\(innbrudd\|språk\)" no.txt
#   grep --line-number " 100* " no.txt
#   echo "hello cat baab dog hello" | grep --line-number "hello\|ba*b"
#
# can also use the empty string ε:
#   echo "hello cat baab dog hello" | python matcher.py "he.l.ε.l.o.bε.a*.b.|"
#   python matcher.py "da.t.a.εma.s.k.i.n.e.r.|. ." no.txt

class NFA:
    def __init__(self, initial, accepting):
        # convenient to only have one accepting state per NFA
        self.initial = initial
        self.accepting = accepting

class State:
    def __init__(self, accept):
        self.accept = accept

        # for not adding duplicates in lists, see add and match functions
        self.last_list = 0

        # the start index on the substring we are potentially matching on
        self.start = 0

        # in a general NFA states may have a lot of out edges,
        # but in our construction from regular expressions, we only need two
        self.outa = None # on form (char, State), or None
        self.outb = None # on form (char, State), or None

def generate_nfa_char(c):
    initial = State(False)
    accepting = State(True)
    initial.outa = (c, accepting)
    return NFA(initial, accepting)

def generate_nfa_split(a, b):
    initial = State(False)
    accepting = State(True)
    initial.outa = ('ε', a.initial)
    initial.outb = ('ε', b.initial)
    a.accepting.outa = ('ε', accepting)
    b.accepting.outa = ('ε', accepting)
    a.accepting.accept = False
    b.accepting.accept = False
    return NFA(initial, accepting)

def generate_nfa_concat(a, b):
    a.accepting.outa = ('ε', b.initial)
    a.accepting.accept = False
    return NFA(a.initial, b.accepting)

def generate_nfa_star(a):
    initial = State(False)
    accepting = State(True)
    initial.outa = ('ε', a.initial)
    initial.outb = ('ε', accepting)
    a.accepting.outa = ('ε', a.initial)
    a.accepting.outb = ('ε', accepting)
    a.accepting.accept = False
    return NFA(initial, accepting)

# its more efficient if we do this with fragments of NFAs instead of
# entire NFAs all the time, see https://swtch.com/~rsc/regexp/regexp1.html
# we follow construction as in slides as i think it is simpler conceptually
def generate_nfa(regexp):
    stack = []
    for c in regexp:
        if c == '|':
            b = stack.pop()
            a = stack.pop()
            stack.append(generate_nfa_split(a, b))
        elif c == '.':
            b = stack.pop()
            a = stack.pop()
            stack.append(generate_nfa_concat(a, b))
        elif c == '*':
            stack.append(generate_nfa_star(stack.pop()))
        else:
            stack.append(generate_nfa_char(c))

    return stack.pop()

def add_if_label_match(out, c, start, new_states, list_id):
    if out:
        (label, state) = out
        if label == c:
            add(new_states, state, start, list_id)

def add(new_states, state, start, list_id):
    if state.last_list == list_id:
        return

    state.last_list = list_id
    state.start = start
    new_states.append(state)
    # handle ε-transitions, aka add epsilon closure
    add_if_label_match(state.outa, 'ε', start, new_states, list_id)
    add_if_label_match(state.outb, 'ε', start, new_states, list_id)

def step(c, states, new_states, list_id):
    for state in states:
        add_if_label_match(state.outa, c, state.start, new_states, list_id)
        add_if_label_match(state.outb, c, state.start, new_states, list_id)

def match(s, nfa, char_set):
    list_id = 0
    states = []
    matches = []
    for i in range(len(s)):
        list_id += 1
        # optimization but not needed: if character not in regexp alphabet,
        # avoid running step function and just reset states manually.
        # typically gives > 2x time performance
        if s[i] not in char_set:
            states = []
            continue

        nfa.initial.start = i
        add(states, nfa.initial, i, list_id)
        new_states = []
        step(s[i], states, new_states, list_id)
        states = new_states
        if nfa.accepting in states:
            matches.append((nfa.accepting.start, i + 1))
            # reset states to avoid overlapping matches
            states = []

    return matches

def print_matches(matches, line, line_nr, colours):
    sys.stdout.write(colours.GREEN + str(line_nr))
    sys.stdout.write(colours.BLUE + ':' + colours.ENDC)
    k = 0
    for (i, j) in matches:
        sys.stdout.write(line[k:i])
        sys.stdout.write(colours.RED + line[i:j] + colours.ENDC)
        k = j

    sys.stdout.write(line[k:])

if __name__ == '__main__':
    regexp = sys.argv[1]
    fs = sys.argv[2:]
    nfa = generate_nfa(regexp)
    # char_set not needed but small optimization, see match function
    char_set = {c for c in regexp if c not in '|.*'}

    # only colour output if written to terminal
    isatty = sys.stdout.isatty()
    class colours:
        RED = '\033[31m' if isatty else ''
        GREEN = '\033[32m' if isatty else ''
        BLUE = '\033[34m' if isatty else ''
        ENDC = '\033[m' if isatty else ''

    line_nr = 1
    for line in fileinput.input(files=fs):
        matches = match(line, nfa, char_set)
        if matches:
            print_matches(matches, line, line_nr, colours)

        line_nr += 1
